### ISSUE TITLE
Add "converter" attribute to @Option

### ISSUE THREAD

**Issue by godzsa** [remkop/picocli#237]
I am writing an app which accepts natural language date and parses it with [Natty](http://natty.joestelmach.com/).

So I'd like to transform the given argument, when parsing the attributes from the command line. I think the `@Option` accepting a lambda function or a converter class would be nice to do the job

**Comment by remkop** [comment]
Are you aware that it is possible to [register custom converters](http://picocli.info/#_custom_type_converters) like you describe?

**Comment by godzsa** [comment]
@remkop can I specify the options I want to use the custom converters with?

I mean these custom converters work globally, right? Or at least they seem to work so for me.

But what if I want to for example define a String -> String converter which takes a string as an argument and appends "f" at the end of the string: e.g.   `--add-f asd -> "asdf". But I don't want every string argument to use this converter, only the --add-f. (I know this is a silly example, I just want to show a use-case)

**Comment by remkop** [comment]
You are right that type converters currently can only be specified on a per-type basis. 

An `@Option`/`@Parameters` attribute accepting a converter class would allow specifying a converter on a per-field basis. This makes sense.

Will you be able to provide a pull request for this?

**Comment by godzsa** [comment]
Yes :)

**Comment by remkop** [comment]
I’ve started to work on this, hope you don’t mind.

**Comment by remkop** [comment]
Note to self: TODO release notes
```
Index: src/test/java/picocli/CommandLineTypeConversionTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/test/java/picocli/CommandLineTypeConversionTest.java	(revision )
+++ src/test/java/picocli/CommandLineTypeConversionTest.java	(revision )
@@ -29,6 +29,7 @@
 import java.sql.Driver;
 import java.sql.Time;
 import java.sql.Timestamp;
+import java.sql.Types;
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
 import java.util.ArrayList;
@@ -514,7 +515,7 @@
     }
 
     @Test
-    public void testCustomConverter() {
+    public void testRegisterCustomConverter() {
         class Glob {
             public final String glob;
             public Glob(String glob) { this.glob = glob; }
@@ -537,6 +538,108 @@
         assertEquals(args[0], app.globField.glob);
     }
 
+    static class MyGlobConverter implements ITypeConverter<MyGlob> {
+        public MyGlob convert(String value) throws Exception { return new MyGlob(value); }
+    }
+    static class MyGlob {
+        public final String glob;
+        public MyGlob(String glob) { this.glob = glob; }
+    }
+    @Test
+    public void testAnnotateCustomConverter() {
+        class App {
+            @Parameters(converter = MyGlobConverter.class)
+            MyGlob globField;
+        }
+        CommandLine commandLine = new CommandLine(new App());
+
+        String[] args = {"a*glob*pattern"};
+        List<CommandLine> parsed = commandLine.parse(args);
+        assertEquals("not empty", 1, parsed.size());
+        assertTrue(parsed.get(0).getCommand() instanceof App);
+        App app = (App) parsed.get(0).getCommand();
+        assertEquals(args[0], app.globField.glob);
+    }
+
+    static class SqlTypeConverter implements ITypeConverter<Integer> {
+        public Integer convert(String value) throws Exception {
+            switch (value) {
+                case "ARRAY"   : return Types.ARRAY;
+                case "BIGINT"  : return Types.BIGINT;
+                case "BINARY"  : return Types.BINARY;
+                case "BIT"     : return Types.BIT;
+                case "BLOB"    : return Types.BLOB;
+                case "BOOLEAN" : return Types.BOOLEAN;
+                case "CHAR"    : return Types.CHAR;
+                case "CLOB"    : return Types.CLOB;
+                default: return Types.OTHER;
+            }
+        }
+    }
+    @Test
+    public void testAnnotatedCustomConverterDoesNotConflictWithExistingType() {
+        class App {
+            @Parameters(index = "0", converter = SqlTypeConverter.class)
+            int sqlTypeParam;
+
+            @Parameters(index = "1")
+            int normalIntParam;
+
+            @Option(names = "-t", converter = SqlTypeConverter.class)
+            int sqlTypeOption;
+
+            @Option(names = "-x")
+            int normalIntOption;
+        }
+        App app = new App();
+        CommandLine commandLine = new CommandLine(app);
+
+        String[] args = {"-x", "1234", "-t", "BLOB", "CLOB", "5678"};
+        commandLine.parse(args);
+        assertEquals(1234, app.normalIntOption);
+        assertEquals(Types.BLOB, app.sqlTypeOption);
+        assertEquals(Types.CLOB, app.sqlTypeParam);
+        assertEquals(5678, app.normalIntParam);
+    }
+
+    static class ErrorConverter implements ITypeConverter<Integer> {
+        public Integer convert(String value) throws Exception {
+            throw new IllegalStateException("bad converter");
+        }
+    }
+    @Test
+    public void testAnnotatedCustomConverterErrorHandling() {
+        class App {
+            @Parameters(converter = ErrorConverter.class)
+            int sqlTypeParam;
+        }
+        CommandLine commandLine = new CommandLine(new App());
+        try {
+            commandLine.parse("anything");
+        } catch (CommandLine.ParameterException ex) {
+            assertEquals("Could not convert 'anything' to int for positional parameter at index 0..* (<sqlTypeParam>): java.lang.IllegalStateException: bad converter", ex.getMessage());
+        }
+    }
+    static class CustomConverter implements ITypeConverter<Integer> {
+        public Integer convert(String value) { return Integer.parseInt(value); }
+    }
+    static class Plus23Converter implements ITypeConverter<Integer> {
+        public Integer convert(String value) { return Integer.parseInt(value) + 23; }
+    }
+    static class Plus23ConverterFactory implements CommandLine.IFactory {
+        public Object create(Class<?> cls) throws Exception { return new Plus23Converter(); }
+    }
+    @Test
+    public void testAnnotatedCustomConverterFactory() {
+        class App {
+            @Parameters(converter = CustomConverter.class)
+            int converted;
+        }
+        App app = new App();
+        CommandLine commandLine = new CommandLine(app, new Plus23ConverterFactory());
+        commandLine.parse("100");
+        assertEquals(123, app.converted);
+    }
     private static class EnumParams {
         @Option(names = "-timeUnit")
         TimeUnit timeUnit;
```

**Comment by remkop** [comment]
This issue has been addressed in picocli 2.2, which is now available on JCenter and Maven Central. 
See the [release notes](https://github.com/remkop/picocli/releases/tag/v2.2.0) for details.
